### ISSUE TITLE
single_component dust attenuation silently forfeits both dust precompute mechanisms (dust_tau_v absent from _EB_ATTEN_FREE_OK)

### ISSUE THREAD

**Issue by suchethac** [suchethac/tengri#2485]
## Summary

A model built with `dust_attenuation={'type': 'single_component', ...}` silently forfeits **both** build-time dust precompute mechanisms and runs the exact dense path for every gradient evaluation. The model stays correct, so nothing fails and no test notices.

The single free parameter of that block is `dust_tau_v`, and `dust_tau_v` appears **zero times** in `src/tengri/forward/sed_model.py` on `origin/main`.

## Measured

Six paper-1 CANDELS configurations, all built with `approx=WavePrecomp()`, inspecting `_dust_band_response_cache` and `_energy_balance_lut_cache` after build:

| cfg | attenuation type | free dust params | dust band response | energy-balance LUT |
|-----|------------------|------------------|--------------------|--------------------|
| I   | two_component | `dust_tau_bc dust_tau_diff` | ON | ON |
| II  | **single_component** | **`dust_tau_v`** | **OFF** | **OFF** |
| III | two_component | `dust_tau_bc dust_tau_diff` | ON | ON |
| IV  | two_component | `dust_tau_bc dust_tau_diff` | ON | ON |
| V   | **single_component** | **`dust_tau_v`** | **OFF** | **OFF** |
| VI  | two_component | `dust_tau_bc dust_tau_diff` | ON | ON |

The single-screen Calzetti/SMC parameterization is one of the most widely used attenuation models in SED fitting, so this is not an edge case.

## Mechanism

`_EB_ATTEN_FREE_OK` (`sed_model.py:9236` on main) is a hand-maintained allowlist:

```python
_EB_ATTEN_FREE_OK = frozenset(
    {"dust_tau_bc", "dust_tau_diff", "dust_eta_balance", "dust_log_L_ir"}
)
```

- **Dust emission band response** (`_dust_emission_band_response`): gates on
  `shape_free = bool(free_dust - self._EB_ATTEN_FREE_OK) or ("redshift" in free)`.
  With `dust_tau_v` free, `shape_free` is True and the response is never built.
- **Energy-balance LUT** (`_energy_balance_lut`): gates on `unsafe_free`, computed against
  `_EB_ATTEN_FREE_OK` and `_EB_EMISSION_PARAMS`. `dust_tau_v` is in neither, so the LUT is skipped.

This fails **safe** (correct, slower), which is the documented intent of the allowlist. The bug is the omission, not the gating direction.

## The two halves need different fixes — do not conflate them

**These are separate mechanisms with separate correctness conditions.** Treating them as one lever is how #1770 cost a measured 4.77x.

1. **Dust emission band response — a genuine allowlist omission.**
   `dust_tau_v` changes how much energy is absorbed (hence the `L_ir` amplitude) but not the Dale+2014 emission template *shape*. The response `R` stays a build-time constant. The existing homogeneity probe (evaluate the emitter at two luminosities, require `sed(L) == L*sed(1)`) independently verifies this, so admitting `dust_tau_v` is guarded by a behavioral check and not by the list alone.

2. **Energy-balance LUT — structurally cannot represent a single screen today.**
   `build_energy_balance_lut` (`components/dust/energy_balance_precompute.py`) takes **required** `tau_bc_grid` and `tau_diff_grid` kwargs and returns `EnergyBalanceLUT` carrying both axes. There is no `tau_v` axis. Simply adding `dust_tau_v` to `_EB_ATTEN_FREE_OK` would let the LUT build on `(tau_bc, tau_diff)` while the runtime varies `tau_v` — turning a fail-safe into a **silent wrong `L_absorbed`**. This half needs either a `tau_v` axis in the builder, or an explicit single-screen → two-component mapping with the unused screen pinned, plus node-exact agreement against the exact integral.

## The deeper issue

The allowlist is a hand-maintained list of blessed parameter names, in a file that twice argues against exactly that:

> *"Luminosity-dependent shapes ... are common in IR SED models, so verify the property rather than maintaining a list of which models have it"* — `sed_model.py` ~9297

> *"Ask the declared cross-component contract (ADR-0009) who consumes `L_ir` rather than hardcoding a list, so a future `L_ir` consumer inherits the LUT instead of silently forfeiting it"* — `sed_model.py` ~9172

A new attenuation parameterization inherits the same silent forfeit. Whatever guard is added must be able to **observe a model that forfeits the optimization** — a test asserting "two_component gets the LUT" passes today and would have passed throughout this bug's life.

## Reproduce

Build any model with `dust_attenuation={'type': 'single_component', 'law': 'calzetti', 'tau_v': Uniform(0.0, 3.0), 'all_params': Fixed(DEFAULT)}`, `dust_emission={'type': 'dale2014'}`, `approx=WavePrecomp()`, then inspect `model._dust_band_response_cache` and `model._energy_balance_lut_cache`. Both are `None`.

## Not claimed

The wall-clock impact is **not** established here. Row I (two_component, both mechanisms ON) has a higher median wall time than row V (single_component, both OFF), so other factors dominate the row ordering. The FLOP cost of the forfeit should be measured on the compiled HLO per mechanism before any speedup is quoted.

**Comment by suchethac** [comment]
## Correction and expansion: the LUT has a *second, earlier* blocker, and the root pattern is broader

Measured on the built chains (`_cached_component_chain`), both configurations built with `approx=WavePrecomp()`:

```
Config II (single_component)
   stellar            StellarSEDComponent           isinstance(DustSEDComponent)=False
   nebular            NebularSEDComponent           isinstance(DustSEDComponent)=False
   dust_attenuation   DustAttenuationSEDComponent   isinstance(DustSEDComponent)=False
   dust_emission      Dale2014IRSEDComponent        isinstance(DustSEDComponent)=False
   igm                IGMSEDComponent               isinstance(DustSEDComponent)=False
  --> EB LUT gate `dust is not None`: False

Config I (two_component)
   dust               DustSEDComponent              isinstance(DustSEDComponent)=True
  --> EB LUT gate `dust is not None`: True
```

`single_component` resolves to **`DustAttenuationSEDComponent`**, a different class, registered under a different component name (`"dust_attenuation"`, not `"dust"`).

So the energy-balance LUT is blocked **three** independent ways for a single-screen model, and the allowlist is only one of them:

1. `dust = next((c for c in chain if isinstance(c, DustSEDComponent)), None)` returns `None` — the gate fails here first, before the allowlist is consulted. This line appears at `sed_model.py:7077`, `:7197` and `:9145`.
2. `dust_tau_v` is in neither `_EB_ATTEN_FREE_OK` nor `_EB_EMISSION_PARAMS`.
3. `build_energy_balance_lut` has no `tau_v` axis (`tau_bc_grid`/`tau_diff_grid` are required kwargs).

Fixing only (2) would therefore change nothing for the LUT, and fixing (1)+(2) without (3) would build a LUT on the wrong axes — a silent wrong `L_absorbed`. Any fix must address all three or none.

The dust-emission band response is unaffected by (1): its gate finds the emitter by `getattr(c, "name", "") == "dust_emission"`, which both configurations have. It is blocked by (2) alone, so that half remains the genuine one-line allowlist omission described above.

## The root pattern

The forward model identifies components by **concrete class** and by **hardcoded name**, rather than by a declared capability. Four instances of one anti-pattern, all in `sed_model.py`:

| site | identifies by |
|---|---|
| `for _emitter in ("xray", "radio")` (~:2974) | a hardcoded two-element tuple of names |
| `isinstance(c, DustSEDComponent)` (:7077, :7197, :9145) | a concrete class |
| `getattr(c, "name", "") == "dust_emission"` (~:9246) | a hardcoded name |
| `_EB_ATTEN_FREE_OK` / `_EB_EMISSION_PARAMS` (:9043, :9080) | hardcoded parameter names (25 of them) |

Each one silently excludes anything not enumerated, and because every exclusion is fail-safe, no test can distinguish "correctly skipped" from "forgotten". The file argues against precisely this twice in its own comments (~:9172, ~:9297): ask the declared contract, and verify the property rather than maintaining a list.

A guard for this must be able to observe a model that *forfeits* the optimization. Asserting that `two_component` gets the LUT passes today and would have passed throughout the life of this bug.

**Comment by suchethac** [comment]
## Measured: the band-response half costs 2.09x in gradient FLOPs

Compiled-HLO cost analysis of `jax.grad(sum(predict_photometry))`, on config II's own SSP grid (`fsps_prsc_c3k_a_chabrier`), 9 CANDELS bands, `single_component` Calzetti with `tau_v` free, `dale2014`, `approx=WavePrecomp()`. The two arms differ only by whether `dust_tau_v` is in the band-response allowlist, so this isolates the forfeit rather than comparing two unrelated models.

```
band response engaged (fixed)    : True
band response declined (pre-fix) : True
gradient FLOPs pre-fix   : 72,936,272
gradient FLOPs post-fix  : 34,905,872
speedup                  : 2.090x   (52.1% of the work removed)
```

FLOPs off the compiled HLO rather than wall-clock: this machine is running production fits and a contended timing would measure the load. **FLOPs are not wall-clock** — op structure (sequential `while` loops vs vectorizable fusions) can move the two apart, as #964 documented for the age kernels. Treat 2.09x as the arithmetic saving, not a promised speedup.

This is the band-response half only (PR #2487). The energy-balance LUT half is separate and unmeasured here.

**Comment by suchethac** [comment]
## Measured: the energy-balance LUT is worth 67x, and the single-screen path gets none of it

Now that #2487 has landed and given the dust-emission band response its own allowlist, the energy-balance LUT can be toggled on its own for the first time. Compiled-HLO gradient cost analysis of `predict_photometry`, config II's SSP grid (`fsps_prsc_c3k_a_chabrier`), 8 bands including PACS/SPIRE/MIPS. Band response confirmed **built in both arms**, so the only difference is the LUT.

**single_component**, toggling `dust_tau_v` in `_EB_ATTEN_FREE_OK` (the PR #2489 change):

```
EB LUT built : 34,847,504 gradient FLOPs
EB LUT off   : 34,847,504     <-- bit-identical
max |rel photometry diff| : 0.000e+00
```

**two_component control**, toggling `dust_tau_bc`/`dust_tau_diff`:

```
EB LUT on  :  1,092,110
EB LUT off : 73,511,536     <-- 67.3x
```

## What this means

The LUT is a real and very large optimization — **67x** on two-component models. The single-screen support in #2489 currently gets **zero** of it.

The cache is populated and the structural test passes, so the *producer* side works: the class gate, the allowlist entry, and the degenerate `tau_bc=[0.0]` / `tau_diff=tau_v` grids are all in place. But identical FLOPs and identical photometry with and without the LUT is the signature of a value that never reaches the evaluated graph. **Nothing consumes it on the single-screen path.**

`DustSEDComponent` (two_component) and `DustAttenuationSEDComponent` (single_component) are different classes, and the LUT read almost certainly lives only on the former.

So #2489 is incomplete rather than wrong, and finishing it is worth doing: paper-1 configurations II and V are both single-screen, and 34.8M -> ~1M on the energy-balance integral is a large change in what those rows cost.

## Method note

Toggling the allowlist is the right isolation *only since #2487*. Before it, the band response read the same `_EB_ATTEN_FREE_OK`, so admitting `dust_tau_v` switched on both mechanisms and any measurement conflated them — the earlier 1.80x figure on the pre-merge branch is confounded for exactly that reason and should not be quoted. The band-response half alone is [2.09x](https://github.com/suchethac/tengri/issues/2485#issuecomment-5801477217).

FLOPs off the compiled HLO rather than wall-clock: the machine is running production fits, and FLOPs are contention-immune. They are also not wall-clock — op structure can move the two apart (#964) — so treat 67x as the arithmetic saving.
